x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java

```diff
@@ -770,7 +773,15 @@ private Expression collectResolvedAndReplace(Expression e, Map<String, List<Func
                 List<Function> list = getList(seen, fName);
                 for (Function seenFunction : list) {
                     if (seenFunction != f && f.arguments().equals(seenFunction.arguments())) {
-                        return seenFunction;
+                        // Special check for COUNT: an already seen COUNT function will be returned only if its DISTINCT property
+                        // matches the one from the unresolved function to be checked. 
+                        if (seenFunction instanceof Count) {
+                            if (((Count) seenFunction).distinct() == ((Count) f).distinct()) {
```

Comment:
Why is this check different than here: https://github.com/elastic/elasticsearch/pull/37254/files#diff-bb55908282831d2f432f4a4650d55521R825 ?
Is it safe to cast `(Count) f` without checking it with `instanceof` ?

Reply:
`list` should contain a list of functions with the same name, so at that point that list should only have `COUNT`s in it.

Reply:
At this point the functions are already resolved, while in https://github.com/elastic/elasticsearch/pull/37254/files#diff-bb55908282831d2f432f4a4650d55521R825 the function to check against the list is still an `UnresolvedFunction` instance. This step here deals with duplicates, as well, and it's basically a protection against making `COUNT(DISTINCT fieldX)` and `COUNT(fieldX)` the same.